File: Angular/src/app/core/models/task.ts
export class Task {
    id: number
    title: string
    description: string
    difficulty: number
    priority: number
    reusable: boolean
    file_url: string
    affectations_count:number
    completed: boolean
    
    constructor(
        title: string,
        description: string,
        difficulty: number,
        priority: number,
        reusable: boolean,
        id: number,
        file_url: string,
        affectations_count: number,
        completed: boolean


    ) {
        this.title = title;
        this.description = description;
        this.difficulty = difficulty;
        this.priority = priority;
        this.reusable = reusable;
        this.id = id;
        this.file_url = file_url;
        this.affectations_count=affectations_count;
        this.completed=completed
        
    }
}

File: Rails/app/models/task.rb
class Task < ApplicationRecord
    validates :title, presence: true
    validates :description, presence: true

    validates :difficulty, numericality: { only_integer: true, greater_than_or_equal_to: 0, less_than_or_equal_to: 5 }
    validates :priority,  numericality: { only_integer: true, greater_than_or_equal_to: 0, less_than_or_equal_to: 5 }

    has_many :affectations, dependent: :destroy

    has_one_attached :file

    def file_url
        Rails.application.routes.url_helpers.rails_blob_path(file, only_path: true) if file.attached?
    end

    def affectations_count
        affectations.count
    end


    def completed
        affectations.exists? && affectations.where(end: nil).none?
    end

    def ongoing
        affectations.where.not(start: nil).exists? &&
        affectations.where(end: nil).exists?
    end


    def awaiting
    affectations.where.not(start: nil).none?
    end
end
